lib/runAll.js

```diff
@@ -102,38 +101,48 @@ https://github.com/okonet/lint-staged#using-js-functions-to-customize-linter-com
     return 'No tasks to run.'
   }
 
-  return new Listr(
-    [
-      {
-        title: 'Stashing changes...',
-        skip: async () => {
-          const hasPSF = await git.hasPartiallyStagedFiles({ cwd: gitDir })
-          if (!hasPSF) {
-            return 'No partially staged files found...'
-          }
-          return false
+  const git = new GitWorkflow(gitDir)
+
+  try {
+    await new Listr(
+      [
+        {
+          title: 'Preparing...',
+          task: () => git.stashBackup()
+        },
+        {
+          title: 'Running tasks...',
+          task: () => new Listr(tasks, { ...listrOptions, concurrent: true, exitOnError: false })
         },
-        task: ctx => {
-          ctx.hasStash = true
-          return git.gitStashSave({ cwd: gitDir })
+        {
+          title: 'Applying unstaged changes...',
+          skip: ctx => ctx.hasErrors && 'Skipped because of errors from tasks',
+          task: () => git.restoreUnstagedChanges()
+        },
+        {
+          title: 'Restoring original state due to errors...',
+          enabled: ctx => ctx.hasErrors,
+          task: () => git.restoreOriginalState()
+        },
+        {
+          title: 'Cleaning up...',
+          task: () => git.dropBackup()
         }
-      },
-      {
-        title: 'Running tasks...',
-        task: () => new Listr(tasks, { ...listrOptions, concurrent: true, exitOnError: false })
-      },
-      {
-        title: 'Updating stash...',
-        enabled: ctx => ctx.hasStash,
-        skip: ctx => ctx.hasErrors && 'Skipping stash update since some tasks exited with errors',
-        task: () => git.updateStash({ cwd: gitDir })
-      },
-      {
-        title: 'Restoring local changes...',
-        enabled: ctx => ctx.hasStash,
-        task: () => git.gitStashPop({ cwd: gitDir })
-      }
-    ],
-    listrOptions
-  ).run()
+      ],
+      listrOptions
+    ).run()
+  } catch (error) {
+    if (error.message.includes('Another git process seems to be running in this repository')) {
+      logger.error(`
+  ${symbols.error} ${chalk.red(`lint-staged failed due to a git error.
+    Any lost modifications can be restored from a git stash:
+
+    > git stash list
+    stash@{0}: On master: automatic lint-staged backup
+    > git stash pop stash@{0}`)}
```

Comment:
Should we use named stashed here?

Reply:
This is the example text. You can see the stash is named as `automatic lint-staged backup`. I opted for the number in case the actual output shows more stashes. The idea is to apply the matching number.

AFAIK one cannot apply stashes by name: https://stackoverflow.com/questions/11269256/how-to-name-and-retrieve-a-stash-by-name-in-git

Reply:
Oh I see! That’s fine then.